sky/shell/android/update_service_android.cc

```diff
@@ -0,0 +1,129 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "base/android/jni_string.h"
+#include "base/bind.h"
+#include "base/files/file_util.h"
+#include "base/files/file_util.h"
+#include "base/logging.h"
+#include "base/threading/worker_pool.h"
+#include "jni/UpdateService_jni.h"
+#include "mojo/data_pipe_utils/data_pipe_utils.h"
+#include "mojo/public/cpp/application/connect.h"
+#include "mojo/public/interfaces/application/service_provider.mojom.h"
+#include "mojo/services/network/public/interfaces/network_service.mojom.h"
+#include "sky/shell/shell.h"
+
+namespace sky {
+namespace shell {
+
+namespace {
+
+// This helper exists to keep the URLLoaderPtr alive until we call the callback.
+// I'm not sure if this is necessary, though.
+static void CallResponseCallback(
+    mojo::URLLoaderPtr loader,
+    base::Callback<void(mojo::URLResponsePtr)> callback,
+    mojo::URLResponsePtr response) {
+  callback.Run(response.Pass());
+}
+
+static void FetchURL(mojo::NetworkService* network_service,
+                     const std::string& url,
+                     base::Callback<void(mojo::URLResponsePtr)> callback) {
+  mojo::URLLoaderPtr loader;
```

Comment:
It makes the lifetimes easier to reason about.  The way you've got things set up, you've got a bunch of things owned by the message loop and some other things owned by a self-deleting object.  Instead, you can make everything owned by the self-deleting "task" object.

That also prepares the code to be able to cancel the update task in the future.  If the URLLoader is owned by the update task and we delete the update task before CallOnFinished is called, then we'll tear down the URLLoader and cancel the outstanding network request.  If the URLLoader is owned by the event loop, there's no way to cancel it.